swarm/src/lib.rs

```diff
@@ -1684,8 +1684,8 @@ impl fmt::Display for ListenError {
             ListenError::Transport(_) => {
                 write!(f, "Listen error: Failed to negotiate transport protocol(s)")
             }
-            ListenError::Denied { .. } => {
-                write!(f, "Listen error")
+            ListenError::Denied { cause } => {
```

Comment:
This cause is already available as part of calling `Error::source`. Including it in the `fmt::Display` as well is kind of an anti-pattern because it leads to double-printing of errors.

If we print it here, we should remove it from `Error::source`. Also, if you use something like `anyhow`, the source would be printed which I personally find the better approach than printing it here because stringifies as "late" as possible.

Reply:
first let me explain the reason for this: we get an error, we print it and get no information at all about this. The implementation of `Display` comes before the one for `Error` (since `Error: Display` ) so it should be up to the `Error` implementation to deal with this differently